PR: https://github.com/odoo/odoo/pull/218622

From: 56c27882c9a0c0e7d0087798dd851ad457f24fd6
From: Saif (segn)
Date: 2025-07-29 08:54:32

Breaking data model changes scores: del:2 + add:0, change matches:
-    expense_outstanding_account_id = fields.Many2one(
-    expense_outstanding_account_id = fields.Many2one(

Total Changes: 49

[IMP] hr_expense: Remove expense outstanding account setting

As default outstanding account were installed in COA in https://github.com/odoo/odoo/commit/8c60d34c63f88f59bdf45de5ce01d8d87d639fac for flows that
require them.

As expense is one of them, we are removing the expense outstanding
account setting and using the default outstanding account defined in
COA(installing them if not found)

task-4930809

closes odoo/odoo#218622

Related: odoo/upgrade#8054
Signed-off-by: Hugo Poncelet (hupo) <hupo@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_expense/models/hr_expense.py
+++ b/addons/hr_expense/models/hr_expense.py
@@ -1751,22 +1751,7 @@ class HrExpense(models.Model):
     def _get_expense_account_destination(self):
         self.ensure_one()
         if self.payment_mode == 'company_account':
-            journal = self.payment_method_line_id.journal_id
-            account_dest = (
-                self.payment_method_line_id.payment_account_id
-                or journal.company_id.expense_outstanding_account_id
-            )
-            if not account_dest:
-                error_msg = _(
-                    "A default outstanding account must be defined in the settings for company-paid expenses. "
-                    "You can alternatively specify one in the Journal for the %(method)s payment method.",
-                    method=self.payment_method_line_id.display_name,
-                )
-                if self.env['res.config.settings'].has_access('write'):
-                    action = self.env.ref('hr_expense.action_hr_expense_configuration')
-                    raise RedirectWarning(error_msg, action=action.id, button_text=_("Go to settings"))
-                else:
-                    raise UserError(error_msg)
+            account_dest = self.payment_method_line_id.payment_account_id or self._get_outstanding_account_id()
         else:
             if not self.employee_id.sudo().work_contact_id:
                 raise UserError(
@@ -1776,6 +1761,24 @@ class HrExpense(models.Model):
             account_dest = partner.property_account_payable_id or partner.parent_id.property_account_payable_id
         return account_dest.id
 
+    def _get_outstanding_account_id(self):
+        account_ref = 'account_journal_payment_debit_account_id' if self.payment_method_line_id.payment_type == 'inbound' else 'account_journal_payment_credit_account_id'
+        chart_template = self.with_context(allowed_company_ids=self.company_id.root_id.ids).env['account.chart.template']
+        outstanding_account = chart_template.ref(account_ref, raise_if_not_found=False)
+        if not outstanding_account:
+            bank_prefix = self.company_id.bank_account_code_prefix
+            template_data = chart_template._get_chart_template_data(self.company_id.chart_template).get('template_data')
+            code_digits = int(template_data.get('code_digits', 6))
+            chart_template._create_outstanding_accounts(self.company_id, bank_prefix, code_digits)
+            outstanding_account = chart_template.ref(account_ref, raise_if_not_found=False)
+        if not outstanding_account.active:
+            raise RedirectWarning(
+                message=_("The account %(name)s (%(code)s) is archived. Activate it to continue", name=outstanding_account.name, code=outstanding_account.code),
+                action=outstanding_account._get_records_action(),
+                button_text=_("Go to Account"),
+            )
+        return outstanding_account
+
     def _creation_message(self):
         if self.env.context.get('from_split_wizard'):
             return _("Expense created from a split.")

--- a/addons/hr_expense/models/res_company.py
+++ b/addons/hr_expense/models/res_company.py
@@ -13,13 +13,6 @@ class ResCompany(models.Model):
         domain="[('type', '=', 'purchase')]",
         help="The company's default journal used when an employee expense is created.",
     )
-    expense_outstanding_account_id = fields.Many2one(
-        "account.account",
-        string="Outstanding Account",
-        check_company=True,
-        domain="[('account_type', '=', 'asset_current'), ('reconcile', '=', True)]",
-        help="The account used to record the outstanding amount of the company expenses.",
-    )
     company_expense_allowed_payment_method_line_ids = fields.Many2many(
         "account.payment.method.line",
         string="Payment methods available for expenses paid by company",

--- a/addons/hr_expense/models/res_config_settings.py
+++ b/addons/hr_expense/models/res_config_settings.py
@@ -23,13 +23,6 @@ class ResConfigSettings(models.TransientModel):
     module_hr_expense_stripe = fields.Boolean(string='Link your stripe issuing account to manage company credit cards for your employees through Odoo')
 
     expense_journal_id = fields.Many2one('account.journal', related='company_id.expense_journal_id', readonly=False, check_company=True, domain="[('type', '=', 'purchase')]")
-    expense_outstanding_account_id = fields.Many2one(
-        comodel_name='account.account',
-        related='company_id.expense_outstanding_account_id',
-        domain="[('account_type', '=', 'asset_current'), ('reconcile', '=', True)]",
-        readonly=False,
-        check_company=True,
-    )
     company_expense_allowed_payment_method_line_ids = fields.Many2many(
         comodel_name='account.payment.method.line',
         check_company=True,
